test/indexer/test_indexer_controller.py

```diff
@@ -0,0 +1,174 @@
+import json
+import logging
+from typing import (
+    List,
+    Sequence,
+)
+
+import boto3
+from chalice.app import (
+    SQSRecord,
+)
+import mock
+from moto import (
+    mock_sqs,
+    mock_sts,
+)
+
+from azul import (
+    config,
+    queues,
+)
+from azul.azulclient import (
+    AzulClient,
+)
+from azul.indexer import (
+    BundleFQID,
+)
+from azul.indexer.document import (
+    Contribution,
+)
+from azul.indexer.index_controller import (
+    IndexController,
+)
+from azul.indexer.index_service import (
+    CataloguedTallies,
+)
+from azul.logging import (
+    configure_test_logging,
+)
+from indexer import (
+    ForcedRefreshIndexService,
+    IndexerTestCase,
+)
+
+log = logging.getLogger(__name__)
+
+
+# noinspection PyPep8Naming
+def setUpModule():
+    configure_test_logging(log)
+
+
+Contributions = Sequence[Contribution]
+
+
+@mock_sts
+@mock_sqs
+class TestIndexController(IndexerTestCase):
+
+    def setUp(self) -> None:
+        super().setUp()
+        self.index_service.create_indices(self.catalog)
+        self.client = AzulClient()
+        self.controller = IndexController()
+        self.queue_manager = queues.Queues(delete=True)
+
+    def tearDown(self):
+        self.index_service.delete_indices(self.catalog)
+        super().tearDown()
+
+    def _create_mock_queues(self):
+        sqs = boto3.resource('sqs')
+        for queue_name in config.all_queue_names:
+            sqs.create_queue(QueueName=queue_name)
+
+    def _mock_sqs_record(self, **body):
+        event_dict = {
+            'body': json.dumps(body),
+            'receiptHandle': 'ThisWasARandomString',
+            'attributes': {'ApproximateReceiveCount': 1}
+        }
+        return SQSRecord(event_dict=event_dict, context='controller_test')
+
+    def test_contribution_action_invalid(self):
+        event = [
+            self._mock_sqs_record(action='banana',
+                                  notification='slug',
+                                  catalog=self.catalog)
+        ]
+        self.assertRaises(AssertionError, self.controller.contribute, event)
+
+    def test_contribution_action_reindex(self):
+        event = [self._mock_sqs_record(action='reindex', prefix='ff')]
+        with mock.patch('azul.azulclient.AzulClient.do_remote_reindex') as mock_reindex:
+            mock_reindex.return_value = True
+            self.controller.contribute(event)
+            self.assertEqual(1, mock_reindex.call_count)
+
+    def test_contribute_and_aggregate(self):
+        """
+        Index two bundles that make contributions to the same project.
```

Comment:
Note that the first line of the docstring is used for diagnostic output by unit test.

Reply:
![image](https://user-images.githubusercontent.com/5143256/92152220-2e5fc000-edd7-11ea-9165-4ef1dd0bcdb1.png)